nixos/matrix-synapse: Allow passing extras, discover extras from config

mweinelt opened the pull request:
With this change we allow the user to configure extras, exposed as
 optional-dependencies on the matrix-synapse package.
 
 The vertical integration between package, user configuration and
 deployment is a huge boon which then allows us to dynamically adapt the
 python environment the service runs in, by inspecting the configuration
 and autodiscovering certain used extras from config paths.
 
 This allows keeping the generic matrix-synapse build more lean, and
 prevents rebuilds when additional extras or plugins are desired, because
 the environment gets composed from the module. It thereby allows most
 everyone to rely on the cached package from hydra for faster rebuilds.

###### Description of changes

<!--
For package updates please link to a changelog or describe changes, this helps your fellow maintainers discover breaking updates.
For new packages please briefly describe the package or provide a link to its homepage.
-->

###### Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- Built on platform(s)
  - [ ] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [23.05 Release Notes (or backporting 22.11 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2305-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [x] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

<!--
To help with the large amounts of pull requests, we would appreciate your
reviews of other pull requests, especially simple package updates. Just leave a
comment describing what you have tested in the relevant package/service.
Reviewing helps to reduce the average time-to-merge for everyone.
Thanks a lot if you do!

List of open PRs: https://github.com/NixOS/nixpkgs/pulls
Reviewing guidelines: https://nixos.org/manual/nixpkgs/unstable/#chap-reviewing-contributions
-->

mweinelt committed:
matrix-synapse: Prune deps, add missing deps, expose extras

Drops extra arguments on the package, as they will be moved over into
the module in a more discoverable way next.

Also reduces overly broad `with`-scoping.

mweinelt committed:
nixos/matrix-synapse: Allow passing extras, discover extras from config

With this change we allow the user to configure extras, exposed as
optional-dependencies on the matrix-synapse package.

The vertical integration between package, user configuration and
deployment is a huge boon which then allows us to dynamically adapt the
python environment the service runs in, by inspecting the configuration
and autodiscovering certain used extras from config paths.

mweinelt commented:
The default log_config relies on the `JournalHandler` and the tests rely on `psycopg2` (by setting `database.name`) and both cases are proof that the environment works fine.

SuperSandro2000 commented on pkgs/servers/matrix-synapse/default.nix:
```diff
@@ -44,47 +46,93 @@ buildPythonApplication rec {
+    oidc = [
+      authlib
+    ];
+    systemd = lib.optionals (lib.meta.availableOn stdenv.hostPlatform systemd) [
+      systemd
+    ];
```
I would rather report this as an assert error on the module level

mweinelt replied:
This is really an edge case and just makes sure that we provide it only, if supported. I don't think I want to invest much more logic here.

SuperSandro2000 replied:
How about extraFeatures?

Ma27 replied:
Good point. Can't we just turn this optionals into an assertion (I'm fine with `throwIf` for instance).

mweinelt replied:
Would it be enough to just remove the conditional on the list? Then accessing it on an unsupported platform would throw by itself.

Ma27 replied:
good point, fine by me!

SuperSandro2000 commented on nixos/modules/services/matrix/synapse.nix:
```diff
@@ -158,12 +174,33 @@ in {
+          # Provide the JournalHandler for the default log_config
+          "systemd"
+        ];
+        example = literalExpression ''
+          [
+            # Provide statistics about caching memory consumption
+            "cache-memory"
+          ]
+        '';
```
The example should repeat the default settings otherwise the default log_config stops working.

Ma27 replied:
I think we should document somewhere what all of these keys mean, i.e. that systemd is relevant for the log configuration. Adding some of these things as example is probably just confusing (also, the comments for the defaults which also appear relevant won't end up in the manual).

mweinelt replied:
Ideally something upstream maintains, but they only have the comments in the [`pyproject.toml`](https://github.com/matrix-org/synapse/blob/a5fb382a29991c8eafcb8c54cdd8c7aab260c237/pyproject.toml#L249-L273) and then some hints sprinkled across the manual.

Ma27 replied:
Considering that the example was updated I guess we can resolve this part. Personally I think that we should just do this extra-effort, at least I can't think of a way to retrieve this information at eval-time without making the manual depend on `pkgs.matrix-synapse` (and perhaps even ifd) and that's actually a useful piece of information, so that's fine for me.

SuperSandro2000 reviewed (commented):
extras is by itself not very descriptive. How about extraDependencies or similar?

mweinelt commented on nixos/doc/manual/release-notes/rl-2305.section.md:
```diff
@@ -157,6 +157,8 @@ In addition to numerous new and upgraded packages, this release has the followin
 
 - conntrack helper autodetection has been removed from kernels 6.0 and up upstream, and an assertion was added to ensure things don't silently stop working. Migrate your configuration to assign helpers explicitly or use an older LTS kernel branch as a temporary workaround.
 
+- The `matrix-synapse` has stopped providing the `enableSystemd` and `enableRedis` arguments. All extras can now instead be configured from [`services.matrix-synapse.extras`](#opt-services.matrix-synapse.extras), though most of them will automatically be discovered and installed.
```
```suggestion
- The `matrix-synapse` package has stopped providing the `enableSystemd` and `enableRedis` arguments. All extras can now instead be configured from [`services.matrix-synapse.extras`](#opt-services.matrix-synapse.extras), though most of them will automatically be discovered and installed.
```

dali99 commented on pkgs/servers/matrix-synapse/default.nix:
I think this is pretty cool, just a few comments

>This allows keeping the generic matrix-synapse build more lean

In this case it seems also borderline useless? Is the motivation here the cli admin tools that come with?
See other comments.

> prevents rebuilds when additional extras or plugins are desired, because
the environment gets composed from the module. It thereby allows most
everyone to rely on the cached package from hydra for faster rebuilds.

This was also already the case though, I don't think many people were building their own synapses.

mweinelt replied:
By lean I mean that we don't include every possible dependency in the core package.

By allowing to make certain features optional we have two ways. Either we append
them to `propagatedBuildInputs`, which will force a rebuild when diverting from the
default package. Or we modify the `PYTHONPATH` and don't incur that cost.

mweinelt commented on nixos/modules/services/matrix/synapse.nix:
```diff
@@ -158,12 +174,33 @@ in {
+        example = literalExpression ''
+          [
+            # Provide statistics about caching memory consumption
+            "cache-memory"
```
No strong opinion. We could make it conditional on some indicator for prometheus usage.

mweinelt replied:
This seems to require a nested lookup for certain listener resources, which makes this a bit annoying to discover. Would like to consider it out of scope for now.

SuperSandro2000 replied:
We can't reliable automatically discover this since prometheus scraping is often done from another machine and can also be configured in multiple, diverse ways.

mweinelt commented on pkgs/servers/matrix-synapse/default.nix:
```diff
@@ -44,47 +46,93 @@ buildPythonApplication rec {
+  ++ twisted.optional-dependencies.tls;
 
-  nativeCheckInputs = [ mock parameterized openssl ];
+  passthru.optional-dependencies = with python3.pkgs; {
```
One way we coulld this more pleasnt would be to write out the environment from the module (`environment.etc."matrix-synapse/environment" = {}`) and source that in the systemd unit. That way at least it becomes reusable.

dali99 replied:
>Allowing users to rely on the cached package is inherently incompatible with including a wrapper for optional dependencies

1. The module could use the non-wrapped version.
1. Building a wrapper is very common and should not be compared with building the whole package

>In fact the module already relies on the PYTHONPATH for installing plugins, that is only provided through the module.

There are barely any plugins and they are not _core_ functionality

>Lastly I don't think running synapse from a terminal window is a valid use case except for debugging. In which case replicating the PYTHONPATH from the systemd unit would be a straight forward way to go.

There are other reasons to want have working binaries without having to set env variables manually. https://nixery.dev/ for example. Or to be compatible with nix sbom tooling.


Having a standard that applications should work, and their runtime dependencies should be managed and represented by the actual application closure is one of the main selling points of nix for me.

mweinelt replied:
> There are other reasons to want have working binaries without having to set env variables manually. nixery.dev for example. Or to be compatible with nix sbom tooling.

I would argue that applicatons that evaluate at the sbom should also take a look at the runtime environment as specified in the systemd unit, because lots of things can be injected there.

Not sure what the use case for ad-hoc docker containers would be. So then I have `matrix-synapse` in docker, what now?

Coming back to the original question of the thread:

> This is pretty cool, but going from ${pkgs.matrix-synapse}/bin/synapse just working to needing some wrapper script to set PYTHONPATH is a negative. Could you showcase how one would use this package without NixOS with this change?

By overwriting the package and extending `propagatedBuildInputs`. The negative here is that you will have to build the package from source and run the test suite.

```
nix-repl> :b matrix-synapse.overridePythonAttrs (oldAttrs: { propagatedBuildInputs = oldAttrs.propagatedBuildInputs ++ oldAttrs.passthru.optional-dependencies.oidc; })
```

> Maybe pkgs.matrix-synapse could be a functioning matrix-synapse (by wrapping or some other way), and then moving all this stuff into python3.pkgs.matrix-synapse could let us keep nixpkgs' way of neatly enclosing runtime dependencies while still keeping this?

The `python3.pkgs` package set is only intended for libraries, not applications. The relevant part here is, that only applications can use `packageOverrides` to match required dependency versions.

The only way to make the package work in the intended scenario is to modify the `propagatedBuildInputs` as demonstrated above, and generate a package that includes the necessary dependencies.

mweinelt replied:
Personally I don't want to invest time into creating a wrapper function. I don't think they're worth the effort, because they are inherently undiscoverable and require separate documentation, that nobody will know where to look for.

The same goes for the `enableFoo` arguments, which I'm dropping in this changeset.

Ma27 replied:
Just in case it wasn't clear, I agree about the `enable*`-arguments. I wanted to point out that I don't think that a big wrapper containing a bunch of optional things is appropriate here, i.e. I'm in favor of your approach.

dali99 replied:
>I would argue that applicatons that evaluate at the sbom should also take a look at the runtime environment as specified in the systemd unit, because lots of things can be injected there.

>Not sure what the use case for ad-hoc docker containers would be. So then I have matrix-synapse in docker, what now?

I am sure you understood that my examples were just that: examples of external tooling that make (fairly reasonable) assumptions. And as a counter to the strawman that "you shouldn't run synapse in a terminal anyway". And thus arguing their specific merits are beside the point.

As for the sboms: making a an sbom of a package and an sbom for a system configuration are two different things. Having to evaulate a nixos module, and extract the (collection of) relevant systemd units vs just giving it a nixpkgs drv are two entirely different things and arguing that the latter is useless is just bad faith.

>The python3.pkgs package set is only intended for libraries, not applications. The relevant part here is, that only applications can use packageOverrides to match required dependency versions.
>The only way to make the package work in the intended scenario is to modify the propagatedBuildInputs as demonstrated above, and generate a package that includes the necessary dependencies. 

Why is it when the wrapping is done in systemd it's fine to put it in PYTHONPATH but when doing it in a shell script that approach suddenly doesn't work? Am I missing something important here?

>That may be true for plain nixpkgs use, but hard disagree for NixOS use. The runtime environment will often make or break an application.

We're discussing a package in nixpkgs, and how to control its runtime environment/dependencies. Just because this package also has a NixOS service does not mean we should just hose all other consumers.

> How do you define core functionality?

`postgres`, `systemd`, `url-preview`, `redis` (workers), `cache-memory`, `user-search`

Im unsure about `jwt` I think that might(!) be involved in some widget auth stuff, but I'm not sure how.

In the (near<sup>matrix time</sup>) future also `oidc` as that will be the only way to autheticate with synapse.

>Where would you draw the line between plugins that are supposed to be part of said package and which aren't? Also, enable* flags to toggle some behavior (especially things you don't want to ship by default) are hard to discover.

I'd be in favor of dropping the `enableSystemd` and `enableRedis` flags.

As for the wrapper package I don't really care how it's done as long as it has reasonable defaults (like the set above).

If the rest of the dependencies are only a few megabytes I'd probably just add them as well to simplify the package.

There's exactly a thriving ecosystem of plugins so it's not really a problem.

[modules](https://matrix-org.github.io/synapse/latest/modules/index.html) however would also need to be considered...

>looking at the changes here this should still be the case

Only in the NixOS service. The package `matrix-synapse` would have no reason to keep existing.

You could just move the entire package to a let block at top of the module and it'd be more accurate. (well, other than breaking overlays).

>Also, enable* flags to toggle some behavior (especially things you don't want to ship by default) are hard to discover

They are, but I'd argue this incantation:
`pkgs.matrix-synapse.overridePythonAttrs (oldAttrs: { propagatedBuildInputs = oldAttrs.propagatedBuildInputs ++ (with oldAttrs.passthru.optional-dependencies; [ postgres systemd url-preview user-search redis cache-memory oidc ]); })"`
is even harder. (Wtf it can flatten a list of lists?)

Ma27 replied:
> systemd

synapse seems supported on platforms w/o systemd support (see discussion above)

> redis

We don't even have worker support in nixpkgs and even if it will be at some point, it won't be the default.

> They are, but I'd argue this incantation:

That shouldn't be needed at all, the optional dependencies only need to be available in PYTHONPATH, also that'll save you a rebuild.

> [modules](https://matrix-org.github.io/synapse/latest/modules/index.html) however would also need to be considered...

Correct, but I think that this is a good first step in the right direction :) 

Personally, I think an option would be to move the `PYTHONPATH = ...` stuff into a wrapper function in `nixpkgs` then it can be re-used in alternative modules (IIRC your initial remark). Defaults should IMHO remain the one @mweinelt proposed, but rebuilds will be cheap then so I don't consider this a big deal.

dali99 replied:
> > systemd
> 
> synapse seems supported on platforms w/o systemd support (see discussion above)

You'd select it by default on platforms which support it

> > redis
> 
> We don't even have worker support in nixpkgs and even if it will be at some point, it won't be the default.

This will change soon, and I'd argue some workers are actually reasonable to set up by default in an opinionated setup. But that is a discussion for another day.

> Defaults should IMHO remain the one @mweinelt proposed, but rebuilds will be cheap then so I don't consider this a big deal.

Alright, but at that point we'd be arguing over a closure size reduction of <30MB with the added risks of upstream not testing the full matrix (heh) and a more complex package.

> Personally, I think an option would be to move the PYTHONPATH = ... stuff into a wrapper function in nixpkgs then it can be re-used in alternative modules (IIRC your initial remark). Defaults should IMHO remain the one @mweinelt proposed, but rebuilds will be cheap then so I don't consider this a big deal.

Right, so I see this as essentially:

```nix
matrix-synapse = { matrix-synapse-unwrapped, extras ? [ "postgres" (if available then "systemd" else null) "url-preview" "user-search" ], ...}: let
  matrix-synapse-unwrapped = { ... }:  python3.pkgs.buildPythonApplication {...};
in stdenv.mkDerivation {
  nativeBuildInputs = [ makeWrapper ];
  ...
  buildPhase = ''
  	for b in ${matrix-synapse-unwrapped}/bin/*; do
		makeWrapper \
			--set PYTHONPATH=${makeSearchPathOutput "lib" cfg.package.python.sitePackages [ extrasEnv ];}
			"$b" "$out/bin"
	done;
  '';
  passthru.unwrapped = matrix-synapse-unwrapped;
};
```

Then the module can override the extras without needing loss of functionality for non-nixos

mweinelt replied:
Looks like makeWrapper isn't the right tool for the job. Probably needs `wrapPython`? Not sure.

```
matrix-synapse-wrapped> building '/nix/store/2528iqizypzbizvqkg70xzgnfwibw462-matrix-synapse-wrapped-1.84.1.drv'
matrix-synapse-wrapped> -r-xr-xr-x 2 nobody nogroup 3.3K Jan  1  1970 /nix/store/gnk2pmcpwwnc4kb6634pmky1ryf16i64-matrix-synapse-1.84.1/bin/export_signing_key
matrix-synapse-wrapped> -r-xr-xr-x 2 nobody nogroup 3.3K Jan  1  1970 /nix/store/gnk2pmcpwwnc4kb6634pmky1ryf16i64-matrix-synapse-1.84.1/bin/generate_config
matrix-synapse-wrapped> -r-xr-xr-x 2 nobody nogroup 3.3K Jan  1  1970 /nix/store/gnk2pmcpwwnc4kb6634pmky1ryf16i64-matrix-synapse-1.84.1/bin/generate_log_config
matrix-synapse-wrapped> -r-xr-xr-x 2 nobody nogroup 3.3K Jan  1  1970 /nix/store/gnk2pmcpwwnc4kb6634pmky1ryf16i64-matrix-synapse-1.84.1/bin/generate_signing_key
matrix-synapse-wrapped> -r-xr-xr-x 2 nobody nogroup 3.3K Jan  1  1970 /nix/store/gnk2pmcpwwnc4kb6634pmky1ryf16i64-matrix-synapse-1.84.1/bin/hash_password
matrix-synapse-wrapped> -r-xr-xr-x 2 nobody nogroup 3.3K Jan  1  1970 /nix/store/gnk2pmcpwwnc4kb6634pmky1ryf16i64-matrix-synapse-1.84.1/bin/register_new_matrix_user
matrix-synapse-wrapped> -r-xr-xr-x 2 nobody nogroup 3.3K Jan  1  1970 /nix/store/gnk2pmcpwwnc4kb6634pmky1ryf16i64-matrix-synapse-1.84.1/bin/synapse_homeserver
matrix-synapse-wrapped> -r-xr-xr-x 2 nobody nogroup 3.3K Jan  1  1970 /nix/store/gnk2pmcpwwnc4kb6634pmky1ryf16i64-matrix-synapse-1.84.1/bin/synapse_port_db
matrix-synapse-wrapped> -r-xr-xr-x 2 nobody nogroup 3.3K Jan  1  1970 /nix/store/gnk2pmcpwwnc4kb6634pmky1ryf16i64-matrix-synapse-1.84.1/bin/synapse_review_recent_signups
matrix-synapse-wrapped> -r-xr-xr-x 2 nobody nogroup 3.3K Jan  1  1970 /nix/store/gnk2pmcpwwnc4kb6634pmky1ryf16i64-matrix-synapse-1.84.1/bin/synapse_worker
matrix-synapse-wrapped> -r-xr-xr-x 2 nobody nogroup 3.3K Jan  1  1970 /nix/store/gnk2pmcpwwnc4kb6634pmky1ryf16i64-matrix-synapse-1.84.1/bin/synctl
matrix-synapse-wrapped> -r-xr-xr-x 2 nobody nogroup 3.3K Jan  1  1970 /nix/store/gnk2pmcpwwnc4kb6634pmky1ryf16i64-matrix-synapse-1.84.1/bin/update_synapse_database
matrix-synapse-wrapped> 
matrix-synapse-wrapped> Builder called die: Cannot wrap '--set' because it is not an executable file
matrix-synapse-wrapped> Backtrace:
matrix-synapse-wrapped> 7 assertExecutable /nix/store/2y2nabl483f2ri15x11b4sv03ds6z15x-make-shell-wrapper-hook/nix-support/setup-hook
matrix-synapse-wrapped> 43 makeShellWrapper /nix/store/2y2nabl483f2ri15x11b4sv03ds6z15x-make-shell-wrapper-hook/nix-support/setup-hook
matrix-synapse-wrapped> 36 makeWrapper /nix/store/2y2nabl483f2ri15x11b4sv03ds6z15x-make-shell-wrapper-hook/nix-support/setup-hook
matrix-synapse-wrapped> 1552 genericBuild /nix/store/0swa08bbslxc9wbfixbcd196aa1117lx-stdenv-linux/setup
matrix-synapse-wrapped> 6 main /nix/store/6xg259477c90a229xwmb53pdfkn6ig3g-default-builder.sh
matrix-synapse-wrapped>

mweinelt replied:
Okay, looks fixed now. PTAL.

Ma27 commented on nixos/modules/services/matrix/synapse.nix:
```diff
@@ -50,6 +45,27 @@ let
+    ++ lib.optional (cfg.settings ? user_directory) "user-search"
+    ++ lib.optional (cfg.settings.url_preview_enabled) "url-preview"
+    ++ lib.optional (cfg.settings.database.name == "psycopg2") "postgres"
+    ++ lib.optional (wantsModule "ldap_auth_provider.LdapAuthProviderModule") "matrix-synapse-ldap3";
```
This would break when configuring LDAP auht like this, correct?

```
      password_providers = [
        { module = "ldap_auth_provider.LdapAuthProvider";
          config = {};
        }
      ]
```

mweinelt replied:
I wasn't aware you could configure it like that. From https://github.com/matrix-org/matrix-synapse-ldap3/ it looks like `modules` is they way to go.

And looks like the described way is the legacy way to setting this up. https://github.com/matrix-org/synapse/blob/a5fb382a29991c8eafcb8c54cdd8c7aab260c237/synapse/handlers/auth.py#L1858

Ma27 replied:
Heh good to know :) 
But anyways, just realized that that's actually not relevant because right now you have to add the LDAP module manually via `plugins` anyways, so nothing to be broken here ;-)

mweinelt replied:
I think we shouldn't expose the `matrix-synapse-ldap3` as an extra. 

1. It blurs the scope with regard to plugins and is really an outlier
2. We ideally want to run all tests, which means passing all extras into the test environment. With `matrix-synapse-ldap3` there is the problem, that it causes an infinite recursion with synapse itself.

Ma27 reviewed (approved):
Working fine for me. Also, :+1: on the overall approach.

nixos-discourse commented:
This pull request has been mentioned on **NixOS Discourse**. There might be relevant details there:

https://discourse.nixos.org/t/breaking-changes-announcement-for-unstable/17574/22

SuperSandro2000 commented on nixos/modules/services/matrix/synapse.nix:
```diff
@@ -158,6 +170,59 @@ in {
+          "user-search"
+        ];
+        example = literalExpression ''
+          [
+            # Provide statistics about caching memory consumption
+            "cache-memory"
+
+            # JSON Web Token authentication
+            "jwt"
+
+            # End-to-end tracing support using Jaeger
+            "opentracing"
+
+            # OpenID Connect authentication
+            "oidc"
+
+            # PostgreSQL database backend
+            "postgresql"
+
+            # Redis support for the replication stream between worker processes
+            "redis"
+
+            # SAML2 authentication
+            "saml2"
+
+            # Error tracking and performance metrics
+            "sentry"
+
+            # Provide the JournalHandler used in the default log_config
+            "systemd"
+
+            # Support for oEmbed URL previews
+            "url-preview"
+
+            # Support internationalized domain names in user-search
+            "user-search"
+          ]
```
```suggestion
          [
            "cache-memory" # Provide statistics about caching memory consumption
            "jwt"          # JSON Web Token authentication
            "opentracing"  # End-to-end tracing support using Jaeger
            "oidc"         # OpenID Connect authentication
            "postgresql"   # PostgreSQL database backend
            "redis"        # Redis support for the replication stream between worker processes
            "saml2"        # SAML2 authentication
            "sentry"       # Error tracking and performance metrics
            "systemd"      # Provide the JournalHandler used in the default log_config
            "url-preview"  # Support for oEmbed URL previews
            "user-search"  # Support internationalized domain names in user-search
          ]
```
we should condense that a bit otherwise it will probably poorly display on search.nixos.org.

mweinelt replied:
I'm not sure how it is going to look on s.n.o.. Do you have other examples in mind, which we can use to get an idea?

SuperSandro2000 replied:
``services.nextcloud.extraApps`` has a longer example, `services.tor.relay.role` has a longer description.

mweinelt commented:
Saving the release notes entry here.

> - The `matrix-synapse` package does not accept the `enableSystemd` and `enableRedis` arguments anymore. All extras can now instead be configured from [`services.matrix-synapse.extras`](#opt-services.matrix-synapse.extras). Most of them will automatically be discovered and installed, when the relevant configuration sections are present.

mweinelt committed:
matrix-synapse: Add wrapper to configure extras

The original matrix-synapse packages is now available with the
-unwrapped prefix.

mweinelt commented on nixos/modules/services/matrix/synapse.nix:
```diff
@@ -151,13 +162,45 @@ in {
+      extras = mkOption {
+        type = types.listOf (types.enum (lib.attrNames cfg.package.optional-dependencies));
+        default = [
+          "postgres"
```
References the `optional-depencies` group `postgres`.

Ma27 commented on nixos/modules/services/matrix/synapse.nix:
```diff
@@ -151,13 +162,45 @@ in {
+          "systemd"
+          "url-preview"
+          "user-search"
+        ];
```
Doesn't that mean that if you add `extras`, the defaults will be discarded?

mweinelt replied:
No, list options are additive. You'd need to `lib.mkForce` to overwrite the defaults.

Ma27 replied:
but
```nix
with import <nixpkgs/lib>;
(evalModules {
  modules = [
    {
      options.kek = mkOption {
        type = types.listOf types.str;
        default = [ "a" "b" ];
      };
    }
    {
      kek = [ "c" ];
    }
  ];
}).config.kek
```

evaluates to

`["c"]`

mweinelt replied:
Ah, so maybe they're `lib.mkDefault`, and adding some overwrites them, but writing from multiple places is additive again.

Ma27 replied:
`lib.mkOptionDefault` (lower prio than `lib.mkDefault`) to be precise.


Personally I'd prefer to list the defaults in `defaultText` and add them in the `config` section. That way it's actually additive and if you want something completely different, you can still use `mkForce`.

Ma27 commented:
LGTM (except for one minor question), but haven't tested it yet.

Also cc @dali99 who originally brought up concerns.

mweinelt committed:
nixos/tests/matrix-synapse: Test redis on postgres instance

This requires the module to pick up on the redis configuration, and add
the required extra packages for redis into the wrapper.

mweinelt commented:
Rebased, squashed and added a test to demonstrate that it picks up the redis dependencies.

Ma27 commented on nixos/modules/services/matrix/synapse.nix:
```diff
@@ -707,6 +752,9 @@ in {
     services.matrix-synapse.configFile = configFile;
 
+    # default them, so they are additive
+    services.matrix-synapse.settings.extras = lib.mkDefault defaultExtras;
```
These would be additive when using `services.matrix-synapse.settings.extras = lib.mkDefault [ "my" "extras" ];`, when omitting this `mkDefault` here, only `my` & `extras` would be set.

Iow the `mkDefault` isn't needed here.

mweinelt replied:
Dropped.

Ma27 commented:
One minor remark, otherwise LGTM and mergeable :+1:

Ma27 commented on nixos/modules/services/matrix/synapse.nix:
```diff
@@ -50,6 +45,29 @@ let
+
+  wrapped = pkgs.matrix-synapse.override {
+    matrix-synapse-unwrapped = cfg.package;
+    extras = wantedExtras;
```
Considering that PYTHONPATH is dropped now, the plugins must be added here as well, correct?

mweinelt replied:
Done in the latest iteration, also updated the wrapper accordingly.

mweinelt commented on pkgs/servers/matrix-synapse/wrapper.nix:
```diff
@@ -0,0 +1,44 @@
+  passthru = {
+    unwrapped = matrix-synapse-unwrapped;
+
+    # for backward compatibility
+    inherit (matrix-synapse-unwrapped) plugins tools;
```
Added so references to `matrix-synapse.{tools,plugins}` don't break.

Ma27 commented on nixos/doc/manual/release-notes/rl-2311.section.md:
```diff
@@ -30,6 +30,8 @@
 
 - `mariadb` now defaults to `mariadb_1011` instead of `mariadb_106`, meaning the default version was upgraded from 10.6.x to 10.11.x. See the [upgrade notes](https://mariadb.com/kb/en/upgrading-from-mariadb-10-6-to-mariadb-10-11/) for potential issues.
 
+-  The `matrix-synapse` package does not accept the `enableSystemd` and `enableRedis` arguments anymore. All extras can now instead be configured from services.matrix-synapse.extras. Most of them will automatically be discovered and installed, when the relevant configuration sections are present.
```
It would be nice to linkify `services.matrix-synapse.extras`. Also, it's probably reasonable to explain a few more implications for those who don't use the module at all (i.e. that `extras` is supposed to be used and that `matrix-synapse` vs `matrix-synapse-unwrapped` are different things).

mweinelt replied:
Expanded, I hope the sentence structure still makes sense now,

Ma27 reviewed (changes requested):
At least the LDAP and mjolnir plugins are failing to build now:

```
python3.10-matrix-synapse-ldap3> Check whether the following modules can be imported: ldap_auth_provider
python3.10-matrix-synapse-ldap3> Traceback (most recent call last):
python3.10-matrix-synapse-ldap3>   File "<string>", line 1, in <module>
python3.10-matrix-synapse-ldap3>   File "<string>", line 1, in <lambda>
python3.10-matrix-synapse-ldap3>   File "/nix/store/lwzzgbnj41d657lpxczk6l5f7d5zcnj1-python3-3.10.11/lib/python3.10/importlib/__init__.py", line 126, in import_module
python3.10-matrix-synapse-ldap3>     return _bootstrap._gcd_import(name[level:], package, level)
python3.10-matrix-synapse-ldap3>   File "<frozen importlib._bootstrap>", line 1050, in _gcd_import
python3.10-matrix-synapse-ldap3>   File "<frozen importlib._bootstrap>", line 1027, in _find_and_load
python3.10-matrix-synapse-ldap3>   File "<frozen importlib._bootstrap>", line 1006, in _find_and_load_unlocked
python3.10-matrix-synapse-ldap3>   File "<frozen importlib._bootstrap>", line 688, in _load_unlocked
python3.10-matrix-synapse-ldap3>   File "<frozen importlib._bootstrap_external>", line 883, in exec_module
python3.10-matrix-synapse-ldap3>   File "<frozen importlib._bootstrap>", line 241, in _call_with_frames_removed
python3.10-matrix-synapse-ldap3>   File "/nix/store/rz4cm4rig1n1l2y04ipwk2bns080a2ky-python3.10-matrix-synapse-ldap3-0.2.2/lib/python3.10/site-packages/ldap_auth_provider.py", line 23, in <module>
python3.10-matrix-synapse-ldap3>     import synapse
python3.10-matrix-synapse-ldap3> ModuleNotFoundError: No module named 'synapse'
error: build of '/nix/store/5rfj8wbvn74qlxgrfd9sysap7n6sjy3z-python3.10-matrix-synapse-ldap3-0.2.2.drv' on 'ssh-ng://builder-ng' failed: builder for '/nix/store/5rfj8wbvn74qlxgrfd9sysap7n6sjy3z-python3.10-matrix-synapse-ldap3-0.2.2.drv' failed with exit code 1
```

mweinelt commented:
> At least the LDAP and mjolnir plugins are failing to build now:
> 
> ```
> python3.10-matrix-synapse-ldap3> Check whether the following modules can be imported: ldap_auth_provider
> python3.10-matrix-synapse-ldap3> Traceback (most recent call last):
> python3.10-matrix-synapse-ldap3>   File "<string>", line 1, in <module>
> python3.10-matrix-synapse-ldap3>   File "<string>", line 1, in <lambda>
> python3.10-matrix-synapse-ldap3>   File "/nix/store/lwzzgbnj41d657lpxczk6l5f7d5zcnj1-python3-3.10.11/lib/python3.10/importlib/__init__.py", line 126, in import_module
> python3.10-matrix-synapse-ldap3>     return _bootstrap._gcd_import(name[level:], package, level)
> python3.10-matrix-synapse-ldap3>   File "<frozen importlib._bootstrap>", line 1050, in _gcd_import
> python3.10-matrix-synapse-ldap3>   File "<frozen importlib._bootstrap>", line 1027, in _find_and_load
> python3.10-matrix-synapse-ldap3>   File "<frozen importlib._bootstrap>", line 1006, in _find_and_load_unlocked
> python3.10-matrix-synapse-ldap3>   File "<frozen importlib._bootstrap>", line 688, in _load_unlocked
> python3.10-matrix-synapse-ldap3>   File "<frozen importlib._bootstrap_external>", line 883, in exec_module
> python3.10-matrix-synapse-ldap3>   File "<frozen importlib._bootstrap>", line 241, in _call_with_frames_removed
> python3.10-matrix-synapse-ldap3>   File "/nix/store/rz4cm4rig1n1l2y04ipwk2bns080a2ky-python3.10-matrix-synapse-ldap3-0.2.2/lib/python3.10/site-packages/ldap_auth_provider.py", line 23, in <module>
> python3.10-matrix-synapse-ldap3>     import synapse
> python3.10-matrix-synapse-ldap3> ModuleNotFoundError: No module named 'synapse'
> error: build of '/nix/store/5rfj8wbvn74qlxgrfd9sysap7n6sjy3z-python3.10-matrix-synapse-ldap3-0.2.2.drv' on 'ssh-ng://builder-ng' failed: builder for '/nix/store/5rfj8wbvn74qlxgrfd9sysap7n6sjy3z-python3.10-matrix-synapse-ldap3-0.2.2.drv' failed with exit code 1
> ```

Fixed by passing the unwrapped package.

Ma27 commented on nixos/modules/services/matrix/synapse.nix:
```diff
@@ -151,13 +170,40 @@ in {
-        default = pkgs.matrix-synapse;
-        defaultText = literalExpression "pkgs.matrix-synapse";
+        default = pkgs.matrix-synapse-unwrapped;
+        defaultText = literalExpression "pkgs.matrix-synapse-unwrapped";
```
Hmm, it just occurred to me that `package` should probably be made read-only and point to `wrapped` for the following reasons:

* if you want to override, there are overlays: a modifyable `package`-option only makes sense for packages like `systemd` where modifying the package in an overlay would cause a gazillion rebuilds.
* then it's still safe to use `config.services.matrix-synapse.package` and have the exact package that's used in `matrix-synapse.service`. When using workers where you want the same synapse for the worker services that's useful for instance.

mweinelt replied:
Applied in 17dcdfe3d266e76a4b04bd7a42f2feb1d2d44794

mweinelt commented:
Result of `nixpkgs-review pr 221318` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>1 package blacklisted:</summary>
  <ul>
    <li>nixos-install-tools</li>
  </ul>
</details>
<details>
  <summary>9 packages built:</summary>
  <ul>
    <li>matrix-synapse</li>
    <li>matrix-synapse-plugins.matrix-synapse-ldap3</li>
    <li>matrix-synapse-plugins.matrix-synapse-ldap3.dist</li>
    <li>matrix-synapse-plugins.matrix-synapse-mjolnir-antispam</li>
    <li>matrix-synapse-plugins.matrix-synapse-mjolnir-antispam.dist</li>
    <li>matrix-synapse-plugins.matrix-synapse-shared-secret-auth</li>
    <li>matrix-synapse-plugins.matrix-synapse-shared-secret-auth.dist</li>
    <li>matrix-synapse-unwrapped</li>
    <li>matrix-synapse-unwrapped.dist</li>
  </ul>
</details>

SuperSandro2000 commented:
> At least the LDAP and mjolnir plugins are failing to build now:

I have encountered the same failure when overriding python3 to python311. No idea why tbh.

Ma27 committed:
nixos/matrix-synapse: expose final matrix-synapse package via `package`-option

When extending this module, it might be necessary to run something
from the package that's used in `matrix-synapse.service` (e.g. for
workers).

Now this can be trivially done by using
`config.services.matrix-synapse.package`. Previously it was necessary to
reuse the `PYTHONPATH` from the environment of `matrix-synapse.service`,
but that one doesn't exist anymore.

Ma27 committed:
nixos/release-notes: reword section for synapse wrapper changes

Ma27 committed:
matrix-synapse: fix PYTHONPATH wrapper

Ma27 committed:
nixos/matrix-synapse: clarify that `extras` are additive

Ma27 committed:
nixos/matrix-synapse: fix path to extras for additive settings

Ma27 committed:
nixos/release-notes: use redis as example for extras in synapse

postgres isn't such a good idea actually because it's added by default
to the wrapper.

Ma27 commented:
also cc @sumnerevans @dali99 @mguentner @Ralith @fadenb any final words for that?

Ralith commented on nixos/modules/services/matrix/synapse.nix:
```diff
@@ -151,10 +169,53 @@ in {
+          However, this means that this option is supposed to be only declared
+          by the `services.matrix-synapse` module itself and is thus read-only.
+          In order to modify `matrix-synapse` itself, use an overlay to override
+          `pkgs.matrix-synapse-unwrapped`.
```
Should we have a `package-unwrapped` option or similar?

Ralith replied:
Writing an overlay feels like a lot more work than setting an option. I don't feel too strongly, though, since I don't see myself ever needing this, short of manually deploying a hotfix or something.

Ma27 replied:
Is `package = pkgs.matrix-synapse-unwrapped.overrideAttrs` that much more than `self: super: { matrix-synapse-unwrapped = super.matrix-synapse-unwrapped.overrideAttrs` though?

SuperSandro2000 replied:
This is rather weird and unaligned with most other modules. Also stacking overlays is more computetional expensive than another option and it breaks people's configs which have overwritten the package in the past.

Ma27 replied:
> This is rather weird and unaligned with most other modules

This `package`-option isn't even standardized to my knowledge and in that case I considered the ability to access the "final" synapse package trivially more important than having _another_ override mechanism.

> Also stacking overlays is more computetional expensive than another option and it breaks people's configs which have overwritten the package in the past.

Do you have any numbers on that? I'm somewhat skeptical about this claim (and haven't seen any benchmark/blogpost/?) on that topic because of Nix's lazy nature the fixpoint computation shouldn't be too expensive. Or is this a direct comparison to `package`?

Also, overlays make it at least possible to combine several changes to a single package on evaluation (i.e. >1 overlay). When you try to do this with `package =`, then it can't be merged by `types.package`.

Ma27 committed:
nixos/matrix-synapse: fix option description of `extras` option

Co-authored-by: Benjamin Saunders <ben.e.saunders@gmail.com>

D4ndellion reviewed (approved):
The diff on this looks really great. Thanks for all your work!

SuperSandro2000 commented on pkgs/servers/matrix-synapse/wrapper.nix:
```diff
@@ -0,0 +1,44 @@
+  searchPath = lib.makeSearchPathOutput "lib" matrix-synapse-unwrapped.python.sitePackages (extraPackages ++ [ pluginsEnv ]);
+in
+stdenv.mkDerivation {
+  name = (lib.appendToName "wrapped" matrix-synapse-unwrapped).name;
```
According to the contributing guide this should have a pname and version.
